### PR TITLE
Dynamic Batch Support for TRT 

### PR DESCRIPTION
**Pull request by codeislife99**
This PR adds support for dynamic batch size for dense, conv , conv_transpose and ops surrounding them

### CODE REVIEW

**Comment by codeislife99** [comment]
cc : @anijain2305 @trevor-m

File: python/tvm/relay/op/contrib/tensorrt.py [hunk 1]
```diff
@@ -198,6 +198,18 @@ def _func_wrapper(expr):
...
             return False
+        if op_name == "multiply":
+            shapes = [
+                [
+                    int(x) if not isinstance(x, tvm.tir.expr.Any) else -1
+                    for x in arg.checked_type.shape
+                ]
+                for arg in args
+            ]
+            if all(
```
Comment (anijain2305) [inline python/tvm/relay/op/contrib/tensorrt.py:209]: Please add a comment why this shape is important

File: python/tvm/relay/op/contrib/tensorrt.py [hunk 2]
```diff
@@ -795,6 +839,38 @@ def conv3d_transpose_annotate_fn(expr):  # pylint: disable=unused-variable
...
+    Visits the Graph recursively and checks if it contains compute heavy ops like convolutions and
+    its transpose, dense and batch mat-mul.
+    """
+
+    def __init__(self):
+        ExprVisitor.__init__(self)
+        self.is_compute_intensive = False
+
+    def visit_call(self, call):
+        heavy_ops = set(
```
Comment (anijain2305) [inline python/tvm/relay/op/contrib/tensorrt.py:853]: s/heavy_ops/compute_intensive_ops/

File: tests/scripts/task_ci_python_setup.sh [hunk 3]
```diff
@@ -31,3 +31,4 @@ set -o pipefail
 echo "Addtiional setup in" ${CI_IMAGE_NAME}
 
 python3 -m pip install --user tlcpack-sphinx-addon==0.1.2 synr==0.2.1
+python3 -m pip install --user torch==1.6.0+cpu torchvision==0.7.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
```
Comment (anijain2305) [inline tests/scripts/task_ci_python_setup.sh:34]: Not sure if we can do this. @zhiics

**Review by anijain2305 — COMMENTED** [review]
Overall LGTM. Please add comments. And I am not sure about the changes for CI.

File: python/tvm/relay/op/contrib/tensorrt.py [hunk 4]
```diff
@@ -607,8 +650,9 @@ def reshape_annotate_fn(expr):  # pylint: disable=unused-variable
         for i, value in enumerate(new_shape):
             if value == -1:
                 new_shape[i] = original_volume // np.prod([x for x in new_shape if x != -1])
+        # Remove batch dimension and see if volumes match
         if shape[0] != new_shape[0]:
-            logger.info("reshape: can't modify batch dimension.")
+            print("reshape: can't modify batch dimension.")
```
Comment (trevor-m) [inline python/tvm/relay/op/contrib/tensorrt.py:655]: Change back to `logger.info`

File: python/tvm/relay/op/contrib/tensorrt.py [hunk 5]
```diff
@@ -808,18 +884,22 @@ def is_valid_subgraph(params, body):
...
-                    input_batch_sizes.append(int(tupe_type.shape[0]))
+                    if not isinstance(tupe_type.shape[0], tvm.tir.expr.Any):
+                        input_batch_sizes.append(int(tupe_type.shape[0]))
             else:
                 # Scalar inputs not allowed
                 if len(var.checked_type.shape) == 0:
                     logger.info("tensorrt: scalar inputs not supported")
                     return False
-                input_batch_sizes.append(int(var.checked_type.shape[0]))
+                if not isinstance(var.checked_type.shape[0], tvm.tir.expr.Any):
```
Comment (trevor-m) [inline python/tvm/relay/op/contrib/tensorrt.py:894]: By not checking the batch size here (and above for tuple type input), we open ourselves up to a possibility where we have partitioned a subgraph for TRT which cannot be executed if the inputs have different batch sizes.

Maybe we should only allow subgraphs with dynamic batch sizes if there is only one input.

This limitation only applies to implicit batch mode in TRT.

File: src/runtime/contrib/tensorrt/tensorrt_runtime.cc [hunk 6]
```diff
@@ -240,30 +249,35 @@ class TensorRTRuntime : public JSONRuntimeBase {
     helper.DeclareField("inputs", &engine_and_context.inputs);
     helper.DeclareField("outputs", &engine_and_context.outputs);
     helper.ReadAllFields(&reader);
-    trt_engine_cache_[symbol_name_] = engine_and_context;
+    const int batch_size = 1;
```
Comment (trevor-m) [inline src/runtime/contrib/tensorrt/tensorrt_runtime.cc:252]: We should probably add the batch size of the engine in the key (file name) of the cached engine.

File: tests/python/contrib/test_tensorrt.py [hunk 7]
```diff
@@ -1034,5 +1039,119 @@ def set_func_attr(func, compile_name, symbol_name):
     tvm.ir.assert_structural_equal(mod_trt, mod_exp, map_free_vars=True)
 
 
+def convert_traced_model_to_vm_trt(
```
Comment (trevor-m) [inline tests/python/contrib/test_tensorrt.py:1042]: I don't think we should have these pytorch tests in the TRT tests.

File: tests/python/contrib/test_tensorrt.py [hunk 8]
```diff
@@ -1034,5 +1039,119 @@ def set_func_attr(func, compile_name, symbol_name):
...
+    shape_list = [(input_name, input_shape)]
+    mod, params = relay.frontend.from_pytorch(traced_module, shape_list)
+    mod, config = tensorrt.partition_for_tensorrt(mod, params, remove_no_mac_subgraphs=True)
+    with tvm.transform.PassContext(opt_level=3, disabled_pass=["FoldScaleAxis"]):
+        vm_trt_exec = relay.vm.compile(mod, target=target, params=params)
+
+    return vm_trt_exec
+
+
+def test_maskrcnn_resnet50() -> None:
```
Comment (trevor-m) [inline tests/python/contrib/test_tensorrt.py:1059]: What properties of this particular model are we trying to test here? Is it really specific to TRT integrattion? Can we make a simple test case which replicates the same behavior instead - i.e. single conv2d op with dynamic batch size in TRT.

**Review by trevor-m — CHANGES_REQUESTED** [review]

File: tests/python/contrib/test_tensorrt.py [hunk 9]
```diff
@@ -1034,5 +1039,119 @@ def set_func_attr(func, compile_name, symbol_name):
...
+            tvm_res[i].asnumpy()[tvm_indices],
+            out[i].numpy()[pt_indices],
+            rtol=tol_val,
+            atol=tol_val,
+        )
+
+
 if __name__ == "__main__":
-    pytest.main([__file__])
+    test_maskrcnn_resnet50()
```
Comment (trevor-m) [inline tests/python/contrib/test_tensorrt.py:1156]: Change back to pytest

**Review by trevor-m — CHANGES_REQUESTED** [review]

File: tests/python/contrib/test_tensorrt.py (hunk 9 above)
Comment (codeislife99) [reply tests/python/contrib/test_tensorrt.py:1156]: oops, thanks !

File: tests/python/contrib/test_tensorrt.py (hunk 8 above)
Comment (codeislife99) [reply tests/python/contrib/test_tensorrt.py:1103]: Resolved offline

File: tests/python/contrib/test_tensorrt.py (hunk 7 above)
Comment (codeislife99) [reply tests/python/contrib/test_tensorrt.py]: Resolved offline

File: python/tvm/relay/op/contrib/tensorrt.py (hunk 4 above)
Comment (codeislife99) [reply python/tvm/relay/op/contrib/tensorrt.py:655]: Done.

File: tests/scripts/task_ci_python_setup.sh (hunk 3 above)
Comment (codeislife99) [reply tests/scripts/task_ci_python_setup.sh]: Removed

File: python/tvm/relay/op/contrib/tensorrt.py (hunk 2 above)
Comment (codeislife99) [reply python/tvm/relay/op/contrib/tensorrt.py:853]: Done.

File: python/tvm/relay/op/contrib/tensorrt.py (hunk 5 above)
Comment (codeislife99) [reply python/tvm/relay/op/contrib/tensorrt.py:897]: As discussed offline, there are usually multiple dynamic inputs to the TRT graph from current OD models. Therefore, I am adding a TODO to resolve this case of different dynamic batch sizes in a future PR.

File: tests/python/contrib/test_tensorrt.py [hunk 10]
```diff
@@ -1034,5 +1038,186 @@ def set_func_attr(func, compile_name, symbol_name):
     tvm.ir.assert_structural_equal(mod_trt, mod_exp, map_free_vars=True)
 
 
+def test_tensorrt_dynamic_batch():
+    if skip_codegen_test():
+        return
+
+    batches_to_test = [1, 1, 2, 3, 1, 3, 2]
```
Comment (anijain2305) [inline tests/python/contrib/test_tensorrt.py:1045]: @trevor-m Should we test batch size of 0 here?

Comment (codeislife99) [reply tests/python/contrib/test_tensorrt.py]: Done.

**Review by trevor-m — APPROVED** [review]
Thanks!

**Review by anijain2305 — APPROVED** [review]

**Comment by anijain2305** [comment]
Thanks @codeislife99 @trevor-m This is merged!

**Pull request merged**